zcash_primitives/src/transaction/txid.rs

```diff
@@ -0,0 +1,296 @@
+    expiry_height: BlockHeight,
+) -> Blake2bHash {
+    let mut h = hasher(ZCASH_HEADERS_HASH_PERSONALIZATION);
+
+    h.write_u32::<LittleEndian>(version.header()).unwrap();
+    h.write_u32::<LittleEndian>(version.version_group_id())
+        .unwrap();
+    h.write_u32::<LittleEndian>(consensus_branch_id.into())
+        .unwrap();
+    h.write_u32::<LittleEndian>(lock_time).unwrap();
+    h.write_u32::<LittleEndian>(expiry_height.into()).unwrap();
+
+    h.finalize()
+}
+
+/// Implements [ZIP 244 section T.2](https://zips.z.cash/zip-0244#t-2-transparent-digest)
+pub(crate) fn hash_transparent_txid_data(txdata: &TransactionData) -> Blake2bHash {
+    let mut h = hasher(ZCASH_TRANSPARENT_HASH_PERSONALIZATION);
+    if !txdata.vin.is_empty() || !txdata.vout.is_empty() {
+        h.write_all(transparent_prevout_hash(&txdata.vin).as_bytes())
+            .unwrap();
+        h.write_all(transparent_sequence_hash(&txdata.vin).as_bytes())
+            .unwrap();
+        h.write_all(transparent_outputs_hash(&txdata.vout).as_bytes())
+            .unwrap();
+    }
+    h.finalize()
+}
+
+/// Implements [ZIP 244 section T.3](https://zips.z.cash/zip-0244#t-3-sapling-digest)
+fn hash_sapling_txid_data(txdata: &TransactionData) -> Blake2bHash {
+    let mut h = hasher(ZCASH_SAPLING_HASH_PERSONALIZATION);
+    if !(txdata.shielded_spends.is_empty() && txdata.shielded_outputs.is_empty()) {
+        h.write_all(hash_sapling_spends(&txdata.shielded_spends).as_bytes())
+            .unwrap();
+
+        h.write_all(hash_sapling_outputs(&txdata.shielded_outputs).as_bytes())
+            .unwrap();
+
+        h.write_all(&txdata.value_balance.to_i64_le_bytes())
+            .unwrap();
+    }
+    h.finalize()
+}
+
+#[cfg(feature = "zfuture")]
+fn hash_tze_txid_data(txdata: &TransactionData) -> Blake2bHash {
+    let mut h = hasher(ZCASH_TZE_HASH_PERSONALIZATION);
+    if !(txdata.tze_inputs.is_empty() && txdata.tze_outputs.is_empty()) {
+        h.write_all(hash_tze_inputs(&txdata.tze_inputs).as_bytes())
+            .unwrap();
+        h.write_all(hash_tze_outputs(&txdata.tze_outputs).as_bytes())
+            .unwrap();
+    }
+    h.finalize()
+}
+
+pub(crate) fn to_hash(
+    consensus_branch_id: BranchId,
+    header_digest: Blake2bHash,
+    transparent_digest: Blake2bHash,
+    sapling_digest: Blake2bHash,
+    #[cfg(feature = "zfuture")] tze_digest: Option<Blake2bHash>,
+) -> Blake2bHash {
+    let mut personal = [0; 16];
+    (&mut personal[..12]).copy_from_slice(ZCASH_TX_PERSONALIZATION_PREFIX);
+    (&mut personal[12..])
+        .write_u32::<LittleEndian>(consensus_branch_id.into())
+        .unwrap();
+
+    let mut h = hasher(&personal);
+    h.write_all(header_digest.as_bytes()).unwrap();
+    h.write_all(transparent_digest.as_bytes()).unwrap();
+    h.write_all(sapling_digest.as_bytes()).unwrap();
+    h.write_all(
+        hasher(ZCASH_ORCHARD_HASH_PERSONALIZATION)
+            .finalize()
+            .as_bytes(),
+    )
+    .unwrap();
```

Comment:
Bug: this is going to result in the wrong txid for v5 transactions containing any Orchard components.

Reply:
Ah, right. So this is where disallowing Orchard components entirely made this possible; without this, we have to backport essentially all the Orchard support, but awkwardly because we can't depend on the `orchard` crate itself.

Reply:
Mmm. Well if we don't backport Orchard components, then mobile SDKs will have a subset of v5 transactions that they can't parse, so we'll need to support re-scanning those later once the wallets upgrade to mobile SDKs with full v5 tx support. In which case, this entire PR is moot, since we can use that same re-scanning logic to handle _all_ v5 txs.